### PR TITLE
WIP: Improvements to make it easier to implement runners other than StandardT

### PR DESCRIPTION
**Pull request by ali-abrar**
* Refactor MonadThunk
* Refactor Nix.Thunk.Basic
* Add Nix.Thunk.Separate
* Add Nix.Fresh.Stable
* Add StableId
* Remove Scopes from Context and replace with ScopeT
* Remove the `t` parameter formerly representing thunks. It can now be inferred from `m`

TODO:
  * [ ] Provide implementations for citations class
  * [ ] Restore Nix.Lint
  * [ ] Fix 3 broken tests
  * [ ] Test thunk refactor more thoroughly (adding new tests as necessary)
  * [ ] Update [Project design intro](https://github.com/haskell-nix/hnix/wiki/Design-of-the-HNix-code-base) in accordance with Thunk changes

### CODE REVIEW

**Comment by Anton-Latukha** [comment]
Hey! Nice to see you!

<details><summary>About the CI</summary>
<p>

We have some CI situations after the difficult situation upstream&downstream together with the epic merge, closed the eyes on the Nix dev env state to merge it.
Now "fixed" the Nix dev env builds & updated the patch to Nixpkgs. But do not bother with that - those are just optional builds.

GHC 8.10 on the other hand falls in the CI because there are currently two mandatory builds - Hackage, Cabal GHC (8.{10,4}). And so because 8.4 falls - 8.10 gets aborted by the CI.

If you want to prune the 8.4 - please start the discussion on that.
It is around time, something around time (4-th major release ago, 3 years ago) to stop supporting 8.4 for these types of projects.
</details>

Nice to see the reduction of the `t`. I often become quite dizzy of these stacks of type variables.

**Comment by Anton-Latukha** [comment]
Can you say, do your Thunk refactor theoretically/practically would be able to address/extend to in the future to receive the depth of provenance? There are: #215 #216 #217 #224 #777, which seems like a basic riffs theme everyone learned.

**Comment by Anton-Latukha** [comment]
Thought the next release would be small.

Seems like the next release would be quite a biggy )

**Comment by Anton-Latukha** [comment]
Nice thing you showed up with the open PR process.

If some parts would become MVP for merge - I am happy to do the multi-merge development process. Chokers are quite hard to merge from the both sides of the coin.

**Comment by ali-abrar** [comment]
Hi Anton. I'll have a look at the thunk issues you mentioned.

There are probably some parts of this that can be separated: perhaps the Context/ScopeT change and the StableId change could be separate PRs.

At the moment, I'm packaging up the example program I was testing these changes against. Hopefully that will help to clarify the reasoning behind some of these changes. I'm also happy to describe them in more detail than I have so far: I just wanted to get this up to get the process started.

I also noticed that README-design.md needs to be updated as a result of these changes.

**Comment by Anton-Latukha** [comment]
At the current point, the rebase onto the `master` should be easy. From the `0.12.0` what I did - basically just removed all `hacky*` and reduced the `principled*` from function names and used NixString Monoid everywhere.

I rebased the PR, but I can not push into it, the edit PR checkmark seems to be turned off.

It is easy to rebase to `master`, it is only a couple of minutes.

**Comment by ali-abrar** [comment]
@Anton-Latukha I caught up with master and added some fixes for GHC8.6.5. CI is starting to look a little better.

**Comment by Anton-Latukha** [comment]
> I rebased the PR locally... It is (currently) easy to rebase to `master`, it is only a couple of minutes.
I talked about rebase the classical way. Onto the branch that merge would be merged into. Classical commit structure allows easier tracking & more automatic rebases to `master` during development.

Man, you do the merges the other way around, please don't shoot the messager - but on big PRs this merges of `master` into PR are destined to fail:
1. Any change in master would break synchronization. Because earlier commits in the chain have a divergent history that is fixed afterward at the top. Changes to the `master` introduce more divergence at the bottom of the PR branch, and since top merges does not cover them - the real `master` state would conflict with top `master` merges in the PR, on which you putting new code, which would require more top merges... That cycle would just tighten.
 
2. And GitHub may "report" that "everything is normal", but GitHub on during the PR merge into the `master` - nor any type of the GitHub automation, nor `git` would not be able to solve the commit (diff) sequencing. `master` and the PR have 2 branches with diverged histories which PR tries to cover at the top.

Since there are 2-3 top merges of `master` state into PR, with the code between them - this definitely currently needs a handcrafted manual rebase onto the branch.

**Comment by Anton-Latukha** [comment]
If you would [allow](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) I can rebase the PR properly. The requirements are: It must have all your current code uploaded. And after the rebase you would need to do the `pull -f`.

I used to do all this stuff also. Good IDE Git interface helped me to learn proper workflows & automate & do the `git` stuff properly. In my case it was/is Magit.

**Comment by ali-abrar** [comment]
@Anton-Latukha the option to allow access doesn't appear on this PR - perhaps because of [this issue](https://github.com/isaacs/github/issues/1681). I've invited you to the fork.

**Comment by Anton-Latukha** [comment]
> @Anton-Latukha the option to allow access doesn't appear on this PR - perhaps because of [this issue](https://github.com/isaacs/github/issues/1681). I've invited you to the fork.

Was guessing that org forks can be in this situation.

Ok. I would make a tea, backup the branch and in an hour would submit a rebase.

**Comment by Anton-Latukha** [comment]
Well, I got distracted a bit & this rebase was more tricky.

The more reverse merges made - the harder the rebase becomes.

All done.

But still can not push:
```
🅸 ~/s/h/hnix:aa/monadthunk-experiment⥄> git push -f obsidian aa/monadthunk-experiment
ERROR: Permission to obsidiansystems/hnix.git denied to Anton-Latukha.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

I do:
```
git remote -v

hSloan	git@github.com:hSloan/hnix.git (fetch)
hSloan	git@github.com:hSloan/hnix.git (push)
layus	git@github.com:layus/hnix.git (fetch)
layus	git@github.com:layus/hnix.git (push)
obsidian	git@github.com:obsidiansystems/hnix.git (fetch)
obsidian	git@github.com:obsidiansystems/hnix.git (push)
```

Just as I did for `layus` and `hSloan` just recently.

So PR editing for PRs from the group repositories seems to do not work at all atm.

**Comment by Anton-Latukha** [comment]
The backup of this PR branch state is made to: https://github.com/haskell-nix/hnix/tree/aa/monadthunk-experiment-backup

Proper rebase uploaded to at: https://github.com/haskell-nix/hnix/tree/monadthunk-experiment-new

You can do this:
```
aa/monadthunk-experiment⥄> git pull -f <hnix_remote> monadthunk-experiment-new
```

That should make everything right.

Further, please use `git rebase -i <hnix_remote> master` to update to the `master` and if something - `git rebase --abort`.
Basically - PRs always do rebases - which are a sync to the main branch. Merge is what the main branch does to merge the patch.

Rebases are much better/easier to do through the good Git interface. I also was doing such things before, because used the CLI, states, situations and Git CLI are complex, streamlining and chunking work in interfaces help a lot. With a good interface, you would navigate complex things in Git in no time.

**Comment by ali-abrar** [comment]
Thanks @Anton-Latukha. Some of the author information was lost during the rebase. I've rebased again to fix that and pushed.

**Comment by ali-abrar** [comment]
And once again for some syntax errors introduced.

**Comment by Anton-Latukha** [comment]
Indeed.

I was quick on a draw to send it. Didn't even checked if it builds at the very end. That is probably good example why I do not like to give time promises, and do them sometimes regardless and this is a good example.

Thanks.

**Comment by Anton-Latukha** [comment]
*(cleaned-up the thread from the old ops stuff)*

**Comment by Anton-Latukha** [comment]
`README-design.md` moved to https://github.com/haskell-nix/hnix/wiki/Design-of-the-HNix-code-base, so when you would get around to it - please provide info on changes there.

File: src/Nix/Builtins.hs [hunk 1]
```diff
@@ -116,12 +120,14 @@ withNixContext mpath action = do
...
+         => m (Scopes m (NValue f m))
 builtins = do
-  ref <- defer $ flip nvSet M.empty <$> buildMap
-  lst <- ([("builtins", ref)] ++) <$> topLevelBuiltins
-  pushScope (M.fromList lst) currentScopes
+  rec lst <- pushScope s $ do
+        ref <- defer $ flip nvSet M.empty <$> buildMap
+        ([("builtins", ref)] ++) <$> topLevelBuiltins
+      let s = M.fromList lst
+  pushScope s currentScopes
```
Comment (layus) [inline src/Nix/Builtins.hs:130]: This is not mere rewriting right. You just fixed the builtins.builtins == builtins issue here. I guess it should make it's way to the release notes ultimately.
@Anton-Latukha what do you think ?

File: src/Nix/Effects.hs [hunk 2]
```diff
@@ -287,3 +292,7 @@ addPath p = either throwError return =<< addToStore (T.pack $ takeFileName p) p
 
 toFile_ :: (Framed e m, MonadStore m) => FilePath -> String -> m StorePath
 toFile_ p contents = addTextToStore (T.pack p) (T.pack contents) HS.empty False
+
+instance MonadStore m => MonadStore (ReaderT r m)
+deriving instance MonadStore m => MonadStore (ScopeT binding r m)
+instance MonadStore m => MonadStore (StateT s m)
```
Comment (layus) [inline src/Nix/Effects.hs:305]: I may be missing something here, but why are these instances needed ?
Would there be a way to make `ScopeT binding r` an instance of `MonadTrans` ?
None of the other monads in this file required this `GeneralizedNewtypeDeriving` + `StandaloneDeriving`. What is special about MonadStore ?

File: src/Nix/Effects/Derivation.hs [hunk 3]
```diff
@@ -288,10 +289,10 @@ defaultDerivationStrict = fromValue @(AttrSet (NValue t f m)) >=> \s -> do
 -- | Build a derivation in a context collecting string contexts.
 -- This is complex from a typing standpoint, but it allows to perform the
 -- full computation without worrying too much about all the string's contexts.
-buildDerivationWithContext :: forall e t f m. (MonadNix e t f m) => AttrSet (NValue t f m) -> WithStringContextT m Derivation
+buildDerivationWithContext :: forall e f m. (MonadNix e f m) => AttrSet (NValue f m) -> WithStringContextT m Derivation
 buildDerivationWithContext drvAttrs = do
     -- Parse name first, so we can add an informative frame
-    drvName     <- getAttr   "name"                      $ extractNixString >=> assertDrvStoreName
+    drvName       <- getAttr "name"                      $ extractNixString >=> assertDrvStoreName
```
Comment (layus) [inline src/Nix/Effects/Derivation.hs]: Nit: The indent was correct, as the whole line is indented less. Not _that_ important :-).

File: src/Nix/Effects/Derivation.hs [hunk 4]
```diff
@@ -338,31 +339,33 @@ buildDerivationWithContext drvAttrs = do
...
-    getAttrOr' :: forall v a. (MonadNix e t f m, FromValue v m (NValue' t f m (NValue t f m)))
+    getAttrOr' :: forall v a. (MonadNix e f m, FromValue v m (NValue' f m (NValue f m)))
       => Text -> m a -> (v -> WithStringContextT m a) -> WithStringContextT m a
     getAttrOr' n d f = case M.lookup n drvAttrs of
       Nothing -> lift d
       Just v  -> withFrame' Info (ErrorCall $ "While evaluating attribute '" ++ show n ++ "'") $
                    fromValue' v >>= f
 
+    getAttrOr :: forall v a. (MonadNix e f m, FromValue v m (NValue' f m (NValue f m)))
+      => Text -> a -> (v -> WithStringContextT m a) -> WithStringContextT m a
```
Comment (layus) [inline src/Nix/Effects/Derivation.hs:361]: Poor you :-/. Having to write these hairy signatures is a pain.

File: src/Nix/Effects/Derivation.hs [hunk 5]
```diff
@@ -5,16 +5,17 @@
 {-# LANGUAGE OverloadedStrings #-}
 {-# LANGUAGE RecordWildCards #-}
 {-# LANGUAGE ScopedTypeVariables #-}
-{-# LANGUAGE TupleSections #-}
 {-# LANGUAGE TypeApplications #-}
- 
+{-# LANGUAGE TypeFamilies #-}
+
```
Comment (layus) [inline src/Nix/Effects/Derivation.hs:11]: Why  are TypeFamilies suddenly required in a lot of places ? It there a simple explanation ?

I guess it is related to the `t` parameter being hidden away in the `m` one ?

File: src/Nix/Thunk.hs [hunk 6]
```diff
@@ -1,61 +1,65 @@
...
-      => m (ThunkId m)
-  freshId = lift freshId
+class MonadTransWrap t where
+  --TODO: Can we enforce that the resulting function is as linear as the provided one?
+  liftWrap :: Monad m => (forall x. m x -> m x) -> t m a -> t m a
+
+instance MonadTransWrap (ReaderT s) where
+  liftWrap f a = do
+    env <- ask
+    lift $ f $ runReaderT a env
```
Comment (layus) [inline src/Nix/Thunk.hs:35]: Am I correct that MonadTransWrap is a simplified MonadBaseControl ? Look very simmilar, with a bit less hassle.

File: tests/TestCommon.hs [hunk 7]
```diff
@@ -36,11 +39,11 @@ hnixEvalFile opts file = do
...
               errorWithoutStackTrace
                 .   show
-                =<< renderFrames @(StdValue (StandardT (StdIdT IO)))
-                      @(StdThunk (StandardT (StdIdT IO)))
+                =<< renderFrames @(StdValue (StandardT IO)) -- (StdValue (StandardT (FreshStableIdT IO)))
+                      @(StdThunk (StandardT IO) IO) -- (StdThunk (StandardT (FreshStableIdT IO)))
                       frames
 
-hnixEvalText :: Options -> Text -> IO (StdValue (StandardT (StdIdT IO)))
+hnixEvalText :: Options -> Text -> IO (NValue Identity (StandardT IO)) -- (StdValue (StandardT (FreshStableIdT IO)))
```
Comment (layus) [inline tests/TestCommon.hs:46]: Reminder to drop the comments in the final iteration :-)

**Review by layus — COMMENTED** [review]
Quite a lot of stuff, with interesting bits flooded into the mass removal of the `t` parameter. The simplification of StandardT is a relief.

Now I am waiting for the next update :-)

File: src/Nix/Builtins.hs (hunk 1 above)
Comment (Anton-Latukha) [reply src/Nix/Builtins.hs:130]: Yes. Definitely, code looks like solves it.
Looked into the code and tried some. Would reserve giving any ideas or active code here, I need a couple of levels to my type programming game to be able to help or give meaningful ideas in the middle of such change.
From type, errors received it already looks pretty close to recurse on builtins in the type code. At this state it looks like Ali probably would just naturally arrive to replace `builtinsBuiltin` with self-referencing `builtins`.

Man, our code is terse. It needs or to understand it all togather with Nix internals, or much more documentation (I talk generally, not to Alis case).

**Comment by Anton-Latukha** [comment]
Current upstream changes to rebase: some class/instance space alignment fixed, `"<CYCLE>"` renamed to `"<expr>"` - that's all.

**Comment by Anton-Latukha** [comment]
<details><summary>A note about rebasing</summary>
<p>

BTW. I now probably have proper access to the PR branch. Because when we first did it, I was trying to rebase before I seen the email with the agreement.
I was careless at the end not only because of the time constraint, I kinda wanted to reach a hand as much as I can, but at the same time to punt to you to do a proper rebase in the end. Because I was under the impression that someone can need to learn to rebase properly. So I just not tested the result at the end and shot from the hip. But also I need to learn from that, to not give short time promises.

If it needs time, please from time to time keep the thread posted. I see you mentioned it.

If something - do not worry about desyncs, we are able to tackle that.

The strategy of tacking depends on the main strategy:

How do you *estimate*: PR "goes long", or "goes short?

If it goes `short` - you currently probably would still be able to figure out rebase yourself.
But soon the rebasing the PR to `master` in one jump would become horrible.
Git is quite bad at doing complex rebases in it.
In Git complex rebases is much better to do incrementally.

Big PRs need a lot of recursive rebasing. Learned it from epic PR of robust POSIX Nix installer that I kept in the most **ideal** for reading it like a guided tour when doing a review and **ideal** merge state for them for 2 years. I'm no longer salty about it. Just took a lot (a week or three) of git groom rebasing to prettify it into an ideal tour guide to merge. It is Nix `#1565`. To be honest, in 2 years they didn't fixed any of the reported bugs in the upstream version of the installer, so I needed to sync only one or twice in two years for something they drive-by hooked into the installer.

If PR goes long
then:
  * Since I know the old `master`, know the whole map (`->`) of what changes were introduced, and know the new `master`.
  * Since the changes I currently introduce are basic (`map -> fmap`, layout, equivalent lambda refactors, reorganizing things in modules and documenting them).
I can do backups as the very first thing, and then after something merged upstream - go and sync those changes into this PR.
And this time I would test them properly also ), so when you return to do work - we would have a current mergeable state in the PR to develop, to discuss, to check with CI and to merge.

Again, it is because the PRs of this size need frequent rebases, otherwise the complexity of the Git rebase increases.
Well, to be honest, one can make the rebase by the commit by commit, but that is the same thing that I propose.

Anyway, the any way you decide it would satisfy me.
</details>

**Comment by Anton-Latukha** [comment]
Backup is at: https://github.com/haskell-nix/hnix/tree/2021-02-08-backup-aa/monadthunk-experiment

**Comment by Anton-Latukha** [comment]
Emacs Magit is great. But sometimes indeed, I from time to time encounter situations where Magit decides to change authorship metadata, probably when the commit hunk requires to make a change to every or to the most of lines, it probably calculates the authorship by line count.

If any changes happen somewhere: it is important to check authorship only once, right before the merge.

**Comment by Anton-Latukha** [comment]
Backup: https://github.com/haskell-nix/hnix/tree/2021-02-09-backup-aa/monadthunk-experiment

**Comment by Anton-Latukha** [comment]
When #805 moved a lot of instances around.

So in this rebase transacted the instances around also.

[Non-transitive map](https://github.com/haskell-nix/hnix/wiki/Design-of-the-HNix-code-base#module-connections-over-the-project), is great in giving the info to what parent to put the type or instances to by picking the top parent in the graph.

**Comment by Anton-Latukha** [comment]
Simply synced PR to `master`.

**Comment by Anton-Latukha** [comment]
Fresh rebase.
The previous version is at: https://github.com/haskell-nix/hnix/tree/2021-02-18-backup-aa/monadthunk-experiment

**Comment by Anton-Latukha** [comment]
Another one.

The previous version is at: https://github.com/haskell-nix/hnix/tree/2021-02-18-2-backup-aa/monadthunk-experiment

**Comment by Anton-Latukha** [comment]
Well. `master` & PR sync gets out of hand.

I can not stop the work in the project, because of this PR.

I do treewide pretty trivial refactors to the source code, doing clean-up, organization, optimization & other stuff.

If to rebase - the `master` changes in the code are documented in the commits & in the ChangeLog.

After the work done in the PR, it would be easier to take the resulting PR diff, and submit it over the new `master` code, for which the `master` triviality and documentation of changes helps tremendously.

Also after refactors in `master` - the logic and the work on it would become much easier. If you would need Kliesly HOFs - I preserve them.

Thank you also very much, HNix definitely needs this work allowing multithreading.

**Comment by Anton-Latukha** [comment]
News:

Think you would find the updated implementation of Standart easier to work with.

Over the project, the code complexity that relates to your work was reduced a bit and received help.

Code now allows equivalently using both normal and old Kleisli-aware implementations that are now `*F`.
So you allowed to mix and match them both during development however it would be simpler to you to simplify the process.

Also new `MonadValue StdValue` function signatures which are annotated, simplified, and straight-forward - should help in development, and if somewhere would be difficult or customization of implementation would help - there are isomorphic old ways for it.

Also in the shortest time, the `Nix.Value` module would go through the same scenario (ordering the arguments into classical positions, the formation of Kleisli into `*F``), and by that our basic `NValue{F,',}` types would receive lawful instances for type classes: https://github.com/haskell-nix/hnix/issues/776.

After that work, renewing, troubleshooting, and merging of the patch should be a relative breeze.

[Sorry for breaking backward compatibility with the current patch code by giving more convenience](https://www.youtube.com/watch?v=7n1ryH3igKc).

**Comment by Anton-Latukha** [comment]
@ali-abrar.

`master` branch desynced with this PR largely because the two-side communication (& so the merge process) stopped, your last message was https://github.com/haskell-nix/hnix/pull/804#issuecomment-756402047 on Jan 7, after that `layus` made a review, I rebased & messaged for ~1.5 months in the thread and got no messages back, so what maintainer can do in that case `¯\_(ツ)_/¯`.

---

Currently, meanwhile from the last post, HNix went through a pretty big refactoring phase.

Сode is cleaned-up & better organized, & more documented, which should simplify the work & understanding of things during work.

The refactoring phase I was doing is done. Currently, I can stop making changes for a month or two, especially to make things happen or merge something like this, we are also precisely at a phase that: what API breakages are needed - they can be merged without remorse.

I am also personally interested in this work. Please, if you want to finish - you can, but write to me so we coordinate efforts. Or would ask at least to supply comprehensive information collected, the idea of it & implementation, on the `scopes`, `provenance` & "`fresh`" thunking (the reason behind `fresh` is still a kind of mystery not only to me but now for John himself also, it seemed to relate to the evaluation/thunk initialization). As I understood the `ScopeT` & thunk mechanism changes are central pieces.
